File: packages/cloudflare/src/tools/positioning.ts
/**
 * Semantic positioning tools for Google Slides.
 *
 * These tools abstract away EMU calculations and transform math, providing
 * intuitive positioning using inches and alignment keywords.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { SlidesClient } from "../api/slides-client.js";
import type { Props } from "../types.js";
import { inchesToEmu, emuToInches } from "../utils/units.js";
import {
  SLIDE_SIZES,
  buildAbsoluteTransform,
  calculateAlignmentPosition,
  extractElementBounds,
  type SlideSize,
} from "../utils/transforms.js";

/**
 * Register positioning tools with the MCP server.
 */
export function registerPositioningTools(
  server: McpServer,
  _env: Env,
  props: Props
): void {
  const client = new SlidesClient({ accessToken: props.accessToken });

  /**
   * position_element - Position and size an element using inches and alignment
   */
  server.tool(
    "position_element",
    "Position and size an element using inches and alignment. You can specify absolute coordinates OR alignment OR both. Alignment is relative to the slide boundaries.",
    {
      presentation_id: z.string().describe("The presentation containing the element"),
      element_id: z.string().describe("The page element to position"),
      x: z.number().optional().describe("X position in inches from left edge"),
      y: z.number().optional().describe("Y position in inches from top edge"),
      width: z.number().optional().describe("New width in inches (null to preserve current)"),
      height: z.number().optional().describe("New height in inches (null to preserve current)"),
      horizontal_align: z.enum(["left", "center", "right"]).optional().describe("Align relative to slide width"),
      vertical_align: z.enum(["top", "center", "bottom"]).optional().describe("Align relative to slide height"),
    },
    async ({ presentation_id, element_id, x, y, width, height, horizontal_align, vertical_align }) => {
      try {
        const presentation = await client.getPresentation(presentation_id);

        // Find the element
        let element: Record<string, unknown> | null = null;
        for (const slide of presentation.slides ?? []) {
          for (const pageElement of slide.pageElements ?? []) {
            if (pageElement.objectId === element_id) {
              element = pageElement as unknown as Record<string, unknown>;
              break;
            }
          }
          if (element) break;
        }

        if (!element) {
          throw new Error(`Element ${element_id} not found in presentation`);
        }

        // Get current bounds
        const currentBounds = extractElementBounds(element as Parameters<typeof extractElementBounds>[0]);

        // Calculate new dimensions
        const newWidth = width !== undefined ? inchesToEmu(width) : currentBounds.width;
        const newHeight = height !== undefined ? inchesToEmu(height) : currentBounds.height;

        // Determine slide size
        let slideSize: SlideSize = SLIDE_SIZES["16:9"];
        const pageSize = presentation.pageSize;
        if (pageSize) {
          const slideSizeWidth = pageSize.width?.magnitude ?? 0;
          const slideSizeHeight = pageSize.height?.magnitude ?? 0;
          if (slideSizeWidth && slideSizeHeight) {
            slideSize = { widthEmu: slideSizeWidth, heightEmu: slideSizeHeight };
          }
        }

        // Calculate position
        let newX: number;
        let newY: number;

        if (horizontal_align !== undefined || vertical_align !== undefined) {
          [newX, newY] = calculateAlignmentPosition(
            slideSize,
            newWidth,
            newHeight,
            horizontal_align,
            vertical_align
          );
          // Override with explicit coordinates if provided
          if (x !== undefined) newX = inchesToEmu(x);
          if (y !== undefined) newY = inchesToEmu(y);
        } else {
          newX = x !== undefined ? inchesToEmu(x) : currentBounds.x;
          newY = y !== undefined ? inchesToEmu(y) : currentBounds.y;
        }

        // Build transform
        const transform = buildAbsoluteTransform(newX, newY);

        // Create update request
        const requests: Record<string, unknown>[] = [
          {
            updatePageElementTransform: {
              objectId: element_id,
              transform,
              applyMode: "ABSOLUTE",
            },
          },
        ];

        // Add size update if needed
        if (width !== undefined || height !== undefined) {
          requests[0] = {
            updatePageElementTransform: {
              objectId: element_id,
              transform: {
                scaleX: currentBounds.width ? newWidth / currentBounds.width : 1,
                scaleY: currentBounds.height ? newHeight / currentBounds.height : 1,
                shearX: 0,
                shearY: 0,
                translateX: newX,
                translateY: newY,
                unit: "EMU",
              },
              applyMode: "ABSOLUTE",
            },
          };
        }

        await client.batchUpdate(presentation_id, requests);

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({
                element_id,
                position: { x_inches: emuToInches(newX), y_inches: emuToInches(newY) },
                size: {
                  width_inches: emuToInches(newWidth),
                  height_inches: emuToInches(newHeight),
                },
              }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );

  /**
   * distribute_elements - Distribute elements evenly across the slide
   */
  server.tool(
    "distribute_elements",
    "Distribute elements evenly across the slide.",
    {
      presentation_id: z.string().describe("The presentation ID"),
      element_ids: z.array(z.string()).describe("Elements to distribute (order matters)"),
      direction: z.enum(["horizontal", "vertical"]).describe("Distribution direction"),
      spacing: z.union([z.number(), z.literal("even")]).default("even").describe("Fixed spacing in inches or 'even' for equal distribution"),
    },
    async ({ presentation_id, element_ids, direction, spacing }) => {
      try {
        const presentation = await client.getPresentation(presentation_id);

        // Get slide size
        let slideSize: SlideSize = SLIDE_SIZES["16:9"];
        const pageSize = presentation.pageSize;
        if (pageSize) {
          const slideSizeWidth = pageSize.width?.magnitude ?? 0;
          const slideSizeHeight = pageSize.height?.magnitude ?? 0;
          if (slideSizeWidth && slideSizeHeight) {
            slideSize = { widthEmu: slideSizeWidth, heightEmu: slideSizeHeight };
          }
        }

        // Collect element info
        const elements: Array<{ id: string; x: number; y: number; width: number; height: number }> = [];
        for (const slide of presentation.slides ?? []) {
          for (const pageElement of slide.pageElements ?? []) {
            if (element_ids.includes(pageElement.objectId ?? "")) {
              const bounds = extractElementBounds(pageElement as unknown as Parameters<typeof extractElementBounds>[0]);
              elements.push({
                id: pageElement.objectId ?? "",
                x: bounds.x,
                y: bounds.y,
                width: bounds.width,
                height: bounds.height,
              });
            }
          }
        }

        // Sort by provided order
        const idOrder = new Map(element_ids.map((id, i) => [id, i]));
        elements.sort((a, b) => (idOrder.get(a.id) ?? 0) - (idOrder.get(b.id) ?? 0));

        if (elements.length < 2) {
          return {
            content: [{ type: "text" as const, text: JSON.stringify({ error: "Need at least 2 elements to distribute" }) }],
            isError: true,
          };
        }

        // Calculate positions
        const requests: Record<string, unknown>[] = [];
        const newPositions: Array<{ element_id: string; x_inches: number; y_inches: number }> = [];

        if (direction === "horizontal") {
          const totalElementWidth = elements.reduce((sum, e) => sum + e.width, 0);
          const gap = spacing === "even"
            ? (slideSize.widthEmu - totalElementWidth) / (elements.length + 1)
            : inchesToEmu(spacing as number);

          let currentX = Math.floor(gap);
          for (const elem of elements) {
            const transform = buildAbsoluteTransform(currentX, elem.y);
            requests.push({
              updatePageElementTransform: {
                objectId: elem.id,
                transform,
                applyMode: "ABSOLUTE",
              },
            });
            newPositions.push({
              element_id: elem.id,
              x_inches: emuToInches(currentX),
              y_inches: emuToInches(elem.y),
            });
            currentX += elem.width + Math.floor(gap);
          }
        } else {
          const totalElementHeight = elements.reduce((sum, e) => sum + e.height, 0);
          const gap = spacing === "even"
            ? (slideSize.heightEmu - totalElementHeight) / (elements.length + 1)
            : inchesToEmu(spacing as number);

          let currentY = Math.floor(gap);
          for (const elem of elements) {
            const transform = buildAbsoluteTransform(elem.x, currentY);
            requests.push({
              updatePageElementTransform: {
                objectId: elem.id,
                transform,
                applyMode: "ABSOLUTE",
              },
            });
            newPositions.push({
              element_id: elem.id,
              x_inches: emuToInches(elem.x),
              y_inches: emuToInches(currentY),
            });
            currentY += elem.height + Math.floor(gap);
          }
        }

        if (requests.length > 0) {
          await client.batchUpdate(presentation_id, requests);
        }

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ elements: newPositions }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );

  /**
   * align_elements - Align multiple elements to each other or to the slide
   */
  server.tool(
    "align_elements",
    "Align multiple elements to each other or to the slide.",
    {
      presentation_id: z.string().describe("The presentation ID"),
      element_ids: z.array(z.string()).describe("Elements to align"),
      alignment: z.enum(["left", "center", "right", "top", "middle", "bottom"]).describe("Edge or center to align"),
      reference: z.enum(["first", "last", "slide"]).default("first").describe("What to align to"),
    },
    async ({ presentation_id, element_ids, alignment, reference }) => {
      try {
        const presentation = await client.getPresentation(presentation_id);

        // Get slide size
        let slideSize: SlideSize = SLIDE_SIZES["16:9"];
        const pageSize = presentation.pageSize;
        if (pageSize) {
          const slideSizeWidth = pageSize.width?.magnitude ?? 0;
          const slideSizeHeight = pageSize.height?.magnitude ?? 0;
          if (slideSizeWidth && slideSizeHeight) {
            slideSize = { widthEmu: slideSizeWidth, heightEmu: slideSizeHeight };
          }
        }

        // Collect element info (preserving order)
        const elements: Array<{ id: string; x: number; y: number; width: number; height: number }> = [];
        for (const elemId of element_ids) {
          for (const slide of presentation.slides ?? []) {
            for (const pageElement of slide.pageElements ?? []) {
              if (pageElement.objectId === elemId) {
                const bounds = extractElementBounds(pageElement as unknown as Parameters<typeof extractElementBounds>[0]);
                elements.push({
                  id: elemId,
                  x: bounds.x,
                  y: bounds.y,
                  width: bounds.width,
                  height: bounds.height,
                });
                break;
              }
            }
          }
        }

        if (elements.length < 1) {
          return {
            content: [{ type: "text" as const, text: JSON.stringify({ error: "No elements found" }) }],
            isError: true,
          };
        }

        // Determine reference element
        let refElem: { x: number; y: number; width: number; height: number };
        if (reference === "first") {
          refElem = elements[0];
        } else if (reference === "last") {
          refElem = elements[elements.length - 1];
        } else {
          refElem = { x: 0, y: 0, width: slideSize.widthEmu, height: slideSize.heightEmu };
        }

        // Calculate target positions
        const requests: Record<string, unknown>[] = [];
        const newPositions: Array<{ element_id: string; x_inches: number; y_inches: number }> = [];

        for (const elem of elements) {
          let newX = elem.x;
          let newY = elem.y;

          if (alignment === "left") {
            newX = refElem.x;
          } else if (alignment === "center") {
            const refCenter = refElem.x + Math.floor(refElem.width / 2);
            newX = refCenter - Math.floor(elem.width / 2);
          } else if (alignment === "right") {
            const refRight = refElem.x + refElem.width;
            newX = refRight - elem.width;
          } else if (alignment === "top") {
            newY = refElem.y;
          } else if (alignment === "middle") {
            const refMiddle = refElem.y + Math.floor(refElem.height / 2);
            newY = refMiddle - Math.floor(elem.height / 2);
          } else if (alignment === "bottom") {
            const refBottom = refElem.y + refElem.height;
            newY = refBottom - elem.height;
          }

          const transform = buildAbsoluteTransform(newX, newY);
          requests.push({
            updatePageElementTransform: {
              objectId: elem.id,
              transform,
              applyMode: "ABSOLUTE",
            },
          });
          newPositions.push({
            element_id: elem.id,
            x_inches: emuToInches(newX),
            y_inches: emuToInches(newY),
          });
        }

        if (requests.length > 0) {
          await client.batchUpdate(presentation_id, requests);
        }

        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ elements: newPositions }, null, 2),
            },
          ],
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text" as const,
              text: `Error: ${error instanceof Error ? error.message : String(error)}`,
            },
          ],
          isError: true,
        };
      }
    }
  );
}

File: packages/python/src/google_slides_mcp/tools/positioning.py
"""Semantic positioning tools for Google Slides.

These tools abstract away EMU calculations and transform math, providing
intuitive positioning using inches and alignment keywords.
"""

from typing import TYPE_CHECKING, Literal

from fastmcp import Context

if TYPE_CHECKING:
    from fastmcp import FastMCP


def register_positioning_tools(mcp: "FastMCP") -> None:
    """Register positioning tools with the MCP application.

    Args:
        mcp: The FastMCP application instance
    """

    @mcp.tool()
    async def position_element(
        ctx: Context,
        presentation_id: str,
        element_id: str,
        x: float | None = None,
        y: float | None = None,
        width: float | None = None,
        height: float | None = None,
        horizontal_align: Literal["left", "center", "right"] | None = None,
        vertical_align: Literal["top", "center", "bottom"] | None = None,
    ) -> dict:
        """Position and size an element using inches and alignment.

        All positions/sizes in INCHES (EMU conversion handled internally).
        Standard slide: 10" x 5.625". You can specify absolute coordinates
        OR alignment OR both.

        Args:
            presentation_id: The presentation containing the element
            element_id: The page element to position
            x: X position in inches from left edge
            y: Y position in inches from top edge
            width: New width in inches (None to preserve current)
            height: New height in inches (None to preserve current)
            horizontal_align: Align relative to slide width
            vertical_align: Align relative to slide height

        Returns:
            Updated element position and size in inches
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.transforms import (
            SLIDE_SIZES,
            build_absolute_transform,
            calculate_alignment_position,
            extract_element_bounds,
        )
        from google_slides_mcp.utils.units import emu_to_inches, inches_to_emu

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        # Get current presentation and element info
        presentation = await service.get_presentation(presentation_id)

        # Find the element in the presentation
        element = None
        for slide in presentation.get("slides", []):
            for page_element in slide.get("pageElements", []):
                if page_element.get("objectId") == element_id:
                    element = page_element
                    break
            if element:
                break

        if not element:
            raise ValueError(f"Element {element_id} not found in presentation")

        # Get current bounds
        current_x, current_y, current_width, current_height = extract_element_bounds(
            element
        )

        # Calculate new dimensions
        new_width = inches_to_emu(width) if width is not None else current_width
        new_height = inches_to_emu(height) if height is not None else current_height

        # Determine slide size (default to 16:9)
        slide_size = SLIDE_SIZES["16:9"]
        page_size = presentation.get("pageSize", {})
        if page_size:
            slide_size_width = page_size.get("width", {}).get("magnitude", 0)
            slide_size_height = page_size.get("height", {}).get("magnitude", 0)
            if slide_size_width and slide_size_height:
                from google_slides_mcp.utils.transforms import SlideSize

                slide_size = SlideSize(int(slide_size_width), int(slide_size_height))

        # Calculate position
        if horizontal_align is not None or vertical_align is not None:
            new_x, new_y = calculate_alignment_position(
                slide_size,
                new_width,
                new_height,
                horizontal=horizontal_align,
                vertical=vertical_align,
            )
            # Override with explicit coordinates if provided
            if x is not None:
                new_x = inches_to_emu(x)
            if y is not None:
                new_y = inches_to_emu(y)
        else:
            new_x = inches_to_emu(x) if x is not None else current_x
            new_y = inches_to_emu(y) if y is not None else current_y

        # Build transform
        transform = build_absolute_transform(new_x, new_y)

        # Create update request
        requests = [
            {
                "updatePageElementTransform": {
                    "objectId": element_id,
                    "transform": transform,
                    "applyMode": "ABSOLUTE",
                }
            }
        ]

        # Add size update if needed
        if width is not None or height is not None:
            requests.append(
                {
                    "updatePageElementTransform": {
                        "objectId": element_id,
                        "transform": {
                            "scaleX": new_width / current_width if current_width else 1,
                            "scaleY": (
                                new_height / current_height if current_height else 1
                            ),
                            "shearX": 0,
                            "shearY": 0,
                            "translateX": new_x,
                            "translateY": new_y,
                            "unit": "EMU",
                        },
                        "applyMode": "ABSOLUTE",
                    }
                }
            )

        await service.batch_update(presentation_id, requests)

        return {
            "element_id": element_id,
            "position": {"x_inches": emu_to_inches(new_x), "y_inches": emu_to_inches(new_y)},
            "size": {
                "width_inches": emu_to_inches(new_width),
                "height_inches": emu_to_inches(new_height),
            },
        }

    @mcp.tool()
    async def distribute_elements(
        ctx: Context,
        presentation_id: str,
        element_ids: list[str],
        direction: Literal["horizontal", "vertical"],
        spacing: float | Literal["even"] = "even",
    ) -> dict:
        """Distribute elements evenly across the slide.

        All measurements in INCHES (EMU conversion handled internally).

        Args:
            presentation_id: The presentation ID
            element_ids: Elements to distribute (order matters)
            direction: Distribution direction
            spacing: Fixed spacing in inches between elements, or "even"
                for equal distribution across the slide

        Returns:
            Dictionary with new positions for each element in inches
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.transforms import (
            SLIDE_SIZES,
            build_absolute_transform,
            extract_element_bounds,
        )
        from google_slides_mcp.utils.units import emu_to_inches, inches_to_emu

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        presentation = await service.get_presentation(presentation_id)

        # Get slide size
        slide_size = SLIDE_SIZES["16:9"]
        page_size = presentation.get("pageSize", {})
        if page_size:
            slide_size_width = page_size.get("width", {}).get("magnitude", 0)
            slide_size_height = page_size.get("height", {}).get("magnitude", 0)
            if slide_size_width and slide_size_height:
                from google_slides_mcp.utils.transforms import SlideSize

                slide_size = SlideSize(int(slide_size_width), int(slide_size_height))

        # Collect element info
        elements = []
        for slide in presentation.get("slides", []):
            for page_element in slide.get("pageElements", []):
                if page_element.get("objectId") in element_ids:
                    bounds = extract_element_bounds(page_element)
                    elements.append(
                        {
                            "id": page_element.get("objectId"),
                            "x": bounds[0],
                            "y": bounds[1],
                            "width": bounds[2],
                            "height": bounds[3],
                        }
                    )

        # Sort by provided order
        id_order = {id_: i for i, id_ in enumerate(element_ids)}
        elements.sort(key=lambda e: id_order.get(e["id"], 0))

        if len(elements) < 2:
            return {"error": "Need at least 2 elements to distribute"}

        # Calculate positions
        requests = []
        new_positions = []

        if direction == "horizontal":
            total_element_width = sum(e["width"] for e in elements)
            if spacing == "even":
                available_space = slide_size.width_emu - total_element_width
                gap = available_space / (len(elements) + 1)
            else:
                gap = inches_to_emu(spacing)

            current_x = int(gap)
            for elem in elements:
                transform = build_absolute_transform(current_x, elem["y"])
                requests.append(
                    {
                        "updatePageElementTransform": {
                            "objectId": elem["id"],
                            "transform": transform,
                            "applyMode": "ABSOLUTE",
                        }
                    }
                )
                new_positions.append(
                    {
                        "element_id": elem["id"],
                        "x_inches": emu_to_inches(current_x),
                        "y_inches": emu_to_inches(elem["y"]),
                    }
                )
                current_x += elem["width"] + int(gap)
        else:  # vertical
            total_element_height = sum(e["height"] for e in elements)
            if spacing == "even":
                available_space = slide_size.height_emu - total_element_height
                gap = available_space / (len(elements) + 1)
            else:
                gap = inches_to_emu(spacing)

            current_y = int(gap)
            for elem in elements:
                transform = build_absolute_transform(elem["x"], current_y)
                requests.append(
                    {
                        "updatePageElementTransform": {
                            "objectId": elem["id"],
                            "transform": transform,
                            "applyMode": "ABSOLUTE",
                        }
                    }
                )
                new_positions.append(
                    {
                        "element_id": elem["id"],
                        "x_inches": emu_to_inches(elem["x"]),
                        "y_inches": emu_to_inches(current_y),
                    }
                )
                current_y += elem["height"] + int(gap)

        if requests:
            await service.batch_update(presentation_id, requests)

        return {"elements": new_positions}

    @mcp.tool()
    async def align_elements(
        ctx: Context,
        presentation_id: str,
        element_ids: list[str],
        alignment: Literal["left", "center", "right", "top", "middle", "bottom"],
        reference: Literal["first", "last", "slide"] = "first",
    ) -> dict:
        """Align multiple elements to each other or to the slide.

        Results returned in INCHES.

        Args:
            element_ids: Elements to align
            alignment: Edge or center to align:
                - left/center/right: Horizontal alignment
                - top/middle/bottom: Vertical alignment
            reference: What to align to:
                - first: Align to the first element
                - last: Align to the last element
                - slide: Align to slide boundaries

        Returns:
            Dictionary with new positions for each element in inches
        """
        from google_slides_mcp.auth.middleware import GoogleAuthMiddleware
        from google_slides_mcp.services.slides_service import SlidesService
        from google_slides_mcp.utils.transforms import (
            SLIDE_SIZES,
            build_absolute_transform,
            extract_element_bounds,
        )
        from google_slides_mcp.utils.units import emu_to_inches

        middleware = GoogleAuthMiddleware()
        credentials = await middleware.extract_credentials(ctx)
        service = SlidesService(credentials)

        presentation = await service.get_presentation(presentation_id)

        # Get slide size
        slide_size = SLIDE_SIZES["16:9"]
        page_size = presentation.get("pageSize", {})
        if page_size:
            slide_size_width = page_size.get("width", {}).get("magnitude", 0)
            slide_size_height = page_size.get("height", {}).get("magnitude", 0)
            if slide_size_width and slide_size_height:
                from google_slides_mcp.utils.transforms import SlideSize

                slide_size = SlideSize(int(slide_size_width), int(slide_size_height))

        # Collect element info (preserving order)
        elements = []
        for elem_id in element_ids:
            for slide in presentation.get("slides", []):
                for page_element in slide.get("pageElements", []):
                    if page_element.get("objectId") == elem_id:
                        bounds = extract_element_bounds(page_element)
                        elements.append(
                            {
                                "id": elem_id,
                                "x": bounds[0],
                                "y": bounds[1],
                                "width": bounds[2],
                                "height": bounds[3],
                            }
                        )
                        break

        if len(elements) < 1:
            return {"error": "No elements found"}

        # Determine reference position
        if reference == "first":
            ref_elem = elements[0]
        elif reference == "last":
            ref_elem = elements[-1]
        else:  # slide
            ref_elem = {
                "x": 0,
                "y": 0,
                "width": slide_size.width_emu,
                "height": slide_size.height_emu,
            }

        # Calculate target position based on alignment
        requests = []
        new_positions = []

        for elem in elements:
            new_x = elem["x"]
            new_y = elem["y"]

            if alignment == "left":
                new_x = ref_elem["x"]
            elif alignment == "center":
                ref_center = ref_elem["x"] + ref_elem["width"] // 2
                new_x = ref_center - elem["width"] // 2
            elif alignment == "right":
                ref_right = ref_elem["x"] + ref_elem["width"]
                new_x = ref_right - elem["width"]
            elif alignment == "top":
                new_y = ref_elem["y"]
            elif alignment == "middle":
                ref_middle = ref_elem["y"] + ref_elem["height"] // 2
                new_y = ref_middle - elem["height"] // 2
            elif alignment == "bottom":
                ref_bottom = ref_elem["y"] + ref_elem["height"]
                new_y = ref_bottom - elem["height"]

            transform = build_absolute_transform(new_x, new_y)
            requests.append(
                {
                    "updatePageElementTransform": {
                        "objectId": elem["id"],
                        "transform": transform,
                        "applyMode": "ABSOLUTE",
                    }
                }
            )
            new_positions.append(
                {
                    "element_id": elem["id"],
                    "x_inches": emu_to_inches(new_x),
                    "y_inches": emu_to_inches(new_y),
                }
            )

        if requests:
            await service.batch_update(presentation_id, requests)

        return {"elements": new_positions}
